### PR TITLE
fix: Cannot set default values in greptimedb cluster

### PR DESCRIPTION
**Pull request by daviderli614**
- Bump **kustomize** version from **v4.5.3** to **v5.3.0**.
- Bump **controller-gen** version from **v0.9.0** to **v0.14.0**.
- Bump **controller-runtime** and some **kubernetes dependencies mod package** to hight version.
- Fix to set default values in greptimedb cluster, we should **SetDefaults()** when every reconciler not only `if len(cluster.Status.ClusterPhase) == 0`:
```
E0830 15:33:31.869706   23888 controller.go:329]  "msg"="Reconciler error" "error"="Service \"basic-meta\" is invalid:  
 [spec.ports[0].port: Invalid value: 0: must be between 1 and 65535, inclusive, spec.ports[0].targetPort: Invalid value: 0:  
 must be between 1 and 65535, inclusive, spec.ports[1].port: Invalid value: 0: must be between 1 and 65535, inclusive,  
 spec.ports[1].targetPort: Invalid value: 0: must be between 1 and 65535, inclusive, spec.ports[1]: Duplicate value:  
 core.ServicePort{Name:\"\", Protocol:\"TCP\", AppProtocol:(*string)(nil), Port:0, TargetPort:intstr.IntOrString{Type:0,  
 IntVal:0, StrVal:\"\"}, NodePort:0}]" "GreptimeDBCluster"={"name":"basic","namespace":"default"}  
 "controller"="greptimedbcluster" "controllerGroup"="greptime.io" "controllerKind"="GreptimeDBCluster" "name"="basic"  
 "namespace"="default" "reconcileID"="132f23b1-06a6-425e-92b0-660a828d9b0a"  
```

## Summary by CodeRabbit

- **New Features**
	- Updated the Custom Resource Definition (CRD) for `greptimedbstandalones`, adding new properties such as `resizePolicy`, `restartPolicy`, and `claims` for improved resource management.
- **Improvements**
	- Enhanced metrics configuration by restructuring how metrics are initialized in the controller manager.
	- Improved type safety by explicitly typing default port variables.
- **Bug Fixes**
	- Refined the logic for setting default values in the cluster reconciliation process to prevent potential errors.
- **Dependency Updates**
	- Upgraded various Go module dependencies to newer versions for improved performance and compatibility.

### CODE REVIEW

**Comment by zyy17** [comment]
@daviderli614 It's not a best practice to fix a bug and upgrade the dependencies simultaneously, which is not about the bug itself.

Can you please make the PR into two standalone PRs:

1. Describe what the bug is and how to reproduce it. Reverting the code logic is not a solution. After reviewing the code, I strongly suspect this is not a bug.

2. Submit another PR to bump the dependencies;

File: Makefile [hunk 1]
```diff
@@ -187,8 +187,8 @@ ENVTEST ?= $(LOCALBIN)/setup-envtest
 GOLANGCI_LINT ?= $(LOCALBIN)/golangci-lint
 
 ## Tool Versions
-KUSTOMIZE_VERSION ?= v4.5.3
-CONTROLLER_TOOLS_VERSION ?= v0.9.0
+KUSTOMIZE_VERSION ?= v5.3.0
```
Comment (zyy17) [inline Makefile:190]: Please don't bump the unrelated dependencies at the same time.

File: apis/v1alpha1/defaulting.go [hunk 2]
```diff
@@ -30,11 +30,11 @@ var (
 	defautlHealthEndpoint = "/health"
 
 	// The default settings for GreptimeDBClusterSpec.
-	defaultHTTPPort       = 4000
-	defaultRPCPort        = 4001
-	defaultMySQLPort      = 4002
-	defaultPostgreSQLPort = 4003
-	defaultMetaRPCPort    = 3002
+	defaultHTTPPort       int32 = 4000
```
Comment (zyy17) [inline apis/v1alpha1/defaulting.go:33]: If we don't specify the type, the default type will be `int`.

File: apis/v1alpha1/greptimedbstandalone_types.go [hunk 3]
```diff
@@ -32,7 +32,7 @@ type GreptimeDBStandaloneSpec struct {
 	TLS *TLSSpec `json:"tls,omitempty"`
 
 	// +optional
-	HTTPServicePort int32 `json:"httpPort,omitempty"`
+	HTTPPort int32 `json:"httpPort,omitempty"`
```
Comment (zyy17) [inline apis/v1alpha1/greptimedbstandalone_types.go:35]: The schema is already `httpPort`, but the variable name is still `HTTPServicePort`. It's not a bug that affects functionality.

File: cmd/operator/app/command.go [hunk 4]
```diff
@@ -66,10 +67,12 @@ func NewOperatorCommand() *cobra.Command {
 
 			mgr, err := ctrl.NewManager(cfg, ctrl.Options{
 				Scheme:                 scheme,
-				MetricsBindAddress:     o.MetricsAddr,
 				HealthProbeBindAddress: o.HealthProbeAddr,
 				LeaderElection:         o.EnableLeaderElection,
 				LeaderElectionID:       leaderElectionID,
+				Metrics: metricsserver.Options{
```
Comment (zyy17) [inline cmd/operator/app/command.go:73]: I think it's an unrelated modification of the bug.

File: apis/v1alpha1/defaulting.go (hunk 2 above)
Comment (zyy17) [reply apis/v1alpha1/defaulting.go:33]: I think it's an unrelated modification of the bug.

File: controllers/greptimedbcluster/controller.go [hunk 5]
```diff
@@ -138,21 +138,14 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
 
+	if err = cluster.SetDefaults(); err != nil {
+		r.Recorder.Event(cluster, corev1.EventTypeWarning, "SetDefaultValuesFailed", fmt.Sprintf("Set default values failed: %v", err))
+		return ctrl.Result{}, err
+	}
+
 	// Means the cluster is just created.
 	if len(cluster.Status.ClusterPhase) == 0 {
 		klog.Infof("Start to create the cluster '%s/%s'", cluster.Namespace, cluster.Name)
-
```
Comment (zyy17) [inline controllers/greptimedbcluster/controller.go:144]: Can you describe the bug and how to reproduce it in the PR? You can also create an issue. Reverting the original logic without any description is not a good practice.

**Comment by daviderli614** [comment]
I will split this PR.

**Pull request closed without merging**